x/evm/keeper/state_transition.go

```diff
@@ -305,6 +309,35 @@ func (k *Keeper) ApplyMessage(evm *vm.EVM, msg core.Message, cfg *params.ChainCo
 	}, nil
 }
 
+// ApplyNativeMessage execute a message on evm from native module
+func (k *Keeper) ApplyNativeMessage(msg core.Message) (*types.MsgEthereumTxResponse, error) {
```

Comment:
can we move some of the duplicated contents of this function to a private function so that this can be reused by `ApplyTransaction`?